src/VisualStudio/Core/Def/Implementation/UnusedReferences/RemoveUnusedReferencesCommandHandler.cs

```diff
@@ -0,0 +1,319 @@
+
+            var project = _workspace.CurrentSolution.GetProject(projectId)!;
+            var unusedReferences = GetUnusedReferencesForProject(project, projectAssetsFile!, targetFrameworkMoniker, cancellationToken);
+
+            return (project, unusedReferences);
+        }
+
+        private ImmutableArray<ReferenceUpdate> GetUnusedReferencesForProject(Project project, string projectAssetsFile, string targetFrameworkMoniker, CancellationToken cancellationToken)
+        {
+            ImmutableArray<ReferenceInfo> unusedReferences = ThreadHelper.JoinableTaskFactory.Run(async () =>
+            {
+                var projectReferences = await _referenceCleanupService.GetProjectReferencesAsync(project.FilePath!, cancellationToken).ConfigureAwait(true);
+                var references = ProjectAssetsReader.ReadReferences(projectReferences, projectAssetsFile, targetFrameworkMoniker);
+
+                return await _unusedReferencesService.GetUnusedReferencesAsync(project, references, cancellationToken).ConfigureAwait(true);
+            });
+
+            var referenceUpdates = unusedReferences
+                .Select(reference => new ReferenceUpdate(reference.TreatAsUsed ? UpdateAction.TreatAsUsed : UpdateAction.Remove, reference))
+                .ToImmutableArray();
+
+            return referenceUpdates;
+        }
+
+        private Window GetUnusedReferencesDialog(Project project, ImmutableArray<ReferenceUpdate> referenceUpdates)
+        {
+            var dialog = _unusedReferenceDialogProvider.CreateDialog(project, referenceUpdates);
+
+            var uiShell = _serviceProvider?.GetService<SVsUIShell, IVsUIShell>();
+            if (uiShell is null)
+            {
+                return dialog;
+            }
+
+            uiShell.GetDialogOwnerHwnd(out var ownerHwnd);
+
+            var windowHelper = new WindowInteropHelper(dialog)
+            {
+                Owner = ownerHwnd
+            };
+
+            uiShell.CenterDialogOnWindow(windowHelper.Handle, IntPtr.Zero);
+
+            return dialog;
+        }
+
+        private void ApplyUnusedReferenceUpdates(Project project, ImmutableArray<ReferenceUpdate> referenceUpdates, CancellationToken cancellationToken)
+        {
+            ThreadHelper.JoinableTaskFactory.Run(
+                () => _unusedReferencesService.UpdateReferencesAsync(project, referenceUpdates, cancellationToken));
+        }
+
+        private static bool TryGetPropertyValue(IVsHierarchy hierarchy, string propertyName, out string? propertyValue)
+        {
+            if (hierarchy is not IVsBuildPropertyStorage storage)
+            {
+                propertyValue = null;
+                return false;
+            }
+
+            return ErrorHandler.Succeeded(storage.GetPropertyValue(propertyName, null, (uint)_PersistStorageType.PST_PROJECT_FILE, out propertyValue));
+        }
+
+        private bool TryGetSelectedProjectHierarchy([NotNullWhen(returnValue: true)] out IVsHierarchy? hierarchy)
```

Comment:
Seems to be duplicate of https://github.com/dotnet/roslyn/blob/7aeb78f800e9144f3485d4e9201c455e7afb6f45/src/VisualStudio/Core/Def/Implementation/Diagnostics/VisualStudioDiagnosticAnalyzerService.cs#L281. Can we move to a shared helper type?

Reply:
Moved to a shared helper.